### ISSUE TITLE
Proposal: Add st.apply() to easily perform in-place operations

### ISSUE THREAD

**Issue by bebosudo** [obspy/obspy#2465]
Hi,
I'm looking at obspy and it seems to be a really useful library for seismology.

I'm trying to work on datasets coming from different sensors.
I want to apply numpy functions on all the traces of a stream.

I came up with the following in order to apply a FFT to a stream, which looks kind of an hack to me:
```python
def apply_fft(trace):
    trace.data = np.fft.fft(trace.data)
    return trace

st_fft = obs.Stream(apply_fft(trace) for trace in st)
```

My question is: is there something in obspy like `apply(func)`, similarly to what provided e.g. by pandas?
It seems that obspy already "delegates" to numpy when a Stream is passed as object to np functions, but this returns naked numpy arrays, so the problem persists:
```python
>>> import obspy, numpy as np
>>> st = obspy.read()
>>> len(np.fft.fft(st))
3
>>> type(np.fft.fft(st))
<class 'numpy.ndarray'>
```

My idea is creating an `apply()` method like `st.apply(np.fft.fft)` or `st.apply(np.sqrt), which changes the stream/trace internal array in-place.
I can work on a pull request if this seems worth to be included in obspy.

**Comment by megies** [comment]
I don't see a big enough benefit to add a new method on Stream which already is quite loaded. We just do..

```
for tr in st:
    tr.data = np.sqrt(tr.data)
```

..which doesn't seem like a huge burden on custom code. But let's hear what other people say. :)

**Comment by megies** [comment]
By the way, glad you're trying out ObsPy, hope it helps you in your workflows :)

**Comment by bebosudo** [comment]
With something like an apply method, you could also chain execution of functions in a more elegant fashion:
```python
st.apply(np.fft.fft).apply(np.sqrt)
```
I think it can be considered an abstraction of other methods like filter, taper, detrend, etc.
No big deal though.

**Comment by megies** [comment]
Like I said, let's hear some other opinions. :-)

**Comment by d-chambers** [comment]
Hi @bebosudo,

I agree with @megies on this one but for different reasons.

First, users need to be aware that manipulating the numpy arrays directly is not safe in that it can break the `Trace` and `Stream` machinery. For example,

```python
import obspy

def slice_two(array):
    """ Slice the array starting at element two until the end of the array """
    return array[2:]

st = obspy.read()
print([tr.stats.starttime for tr in st])

for tr in st:
    tr.data = slice_two(tr.data)

# now the starttime is a lie because we changed the array data. 
print([tr.stats.starttime for tr in st])  
```
However, if one uses the `Trace` and `Stream` methods directly, it *should* not be possible to produce a `Stream`/`Trace` in an invalid state. If we added the `apply` method as you suggest we open the flood gates and break that contract. The way it is now it is a bit more painful to modify the underlying data and therefore more obvious that, in doing so, you are on your own. 

Second, in parity with pandas, if we were to add an `apply` method it should expect a function which takes a `Trace` object, not a numpy array, as its only argument (`DataFrame`:`Stream` :: `Series`:`Trace`). This is because `df.apply` expects a function which takes a single series, not a numpy array (which is also pandas' underlying data structure). However, I would need to think more about whether this is a good idea. 

Two side notes:

You can already chain trace/stream methods together. Just be careful not to depend on intermediate results as, unlike pandas, obspy tends to change the data inplace. I use statements like the following:
```python
import obspy

st = (
    obspy.read().
    detrend('linear').
    filter('highpass', freq=1).
    trim(starttime=obspy.UTCDateTime('2009-08-24T00:20:06'))
)
```
If you are interested in the fusion between obspy and pandas checkout [obsplus](https://github.com/niosh-mining/obsplus), an obspy utility package that is exploring that space.  

And lastly, thank you for taking the time to make a proposal. I  am happy that you are using obspy and thinking about how to make it better. We can always use more contributors so please keep it up!

**Comment by bebosudo** [comment]
Thank you @d-chambers and @megies, it has been an interesting conversation, and thank you for the warm welcome :)

I agree with Derrick on the "breaking of the contract"; I noticed with my script that the content of a trace may no longer be a "valid" trace object, e.g. when applying a FFT on data.

So I agree, this feature is not useful as I proposed it; thank you again for your time :)
I'll give a look at obsplus!
